src/ManageCourses.Api/Middleware/BearerTokenApiKeyHandler.cs

```diff
@@ -0,0 +1,57 @@
+using System.Net.Http;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authentication;
+using System.Text.Encodings.Web;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace GovUk.Education.ManageCourses.Api.Middleware
+{
+    public class BearerTokenApiKeyHandler : AuthenticationHandler<BearerTokenApiKeyOptions>
+    {
+        public BearerTokenApiKeyHandler(IOptionsMonitor<BearerTokenApiKeyOptions> options, ILoggerFactory logger, UrlEncoder encoder, ISystemClock clock) : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+        {
+            var accessToken = Request.GetAccessToken();
+
+            if (!string.IsNullOrEmpty(accessToken))
+            {
+                if (accessToken.Equals(Options.ApiKey))
+                {
+                    var identity = new ClaimsIdentity(
+                    new[] {
+                            new Claim (ClaimTypes.NameIdentifier, "System")
+                    }, BearerTokenApiKeyDefaults.AuthenticationScheme);
+
+                    var princical = new ClaimsPrincipal(identity);
+
+                    var ticket = new AuthenticationTicket(princical, BearerTokenApiKeyDefaults.AuthenticationScheme);
+
+                    return AuthenticateResult.Success(ticket);
+                }
+                else {
+                    return AuthenticateResult.Fail($"Invalid api key: {accessToken}");
+                }
+            }
+
+            return AuthenticateResult.NoResult();
+        }
+
+        protected override Task HandleChallengeAsync(AuthenticationProperties properties)
+        {
+            var authResult = this.HandleAuthenticateOnceSafeAsync().Result;
+            if (!authResult.Succeeded && authResult.Failure != null)
+            {
+                Logger.LogDebug(authResult.Failure, "Failed challenge");
```

Comment:
Why do you log rather than throwing an exception here? If I understand the code correctly this isn't a normal execution path for either a client with the correct or incorrect api key.

Reply:
Exception are expensive and it needs to return expected behaviour to state the authentication result

ie a bad password is normal behaviour.

The log is there to catch exception and messages
ie
`return AuthenticateResult.Fail($"Invalid api key: {accessToken}")`

Reply:
Do you not want to know what values was attempted and failed?

Reply:
Okay, I defer to your superior knowledge on the auth stuff.
I think the accessToken info should go, yes. If we find it hard to troubleshoot then we can think of another way of getting it.

Reply:
That's why it's in debug, it's a value that was not expected.